core/engine/src/value/mod.rs

```diff
@@ -63,38 +65,79 @@ static TWO_E_63: Lazy<BigInt> = Lazy::new(|| {
     BigInt::from(TWO_E_63)
 });
 
-/// A Javascript value
-#[derive(Finalize, Debug, Clone)]
-pub enum JsValue {
-    /// `null` - A null value, for when a value doesn't exist.
+/// The Inner type of a [`JsValue`]. This is the actual value that the `JsValue` holds.
+/// This is not a public API and should not be used directly.
+///
+/// If you need access to the variant, use [`JsValue::variant`] instead.
+#[derive(Finalize, Debug, Clone, PartialEq)]
+enum InnerValue {
     Null,
-    /// `undefined` - An undefined value, for when a field or index doesn't exist.
     Undefined,
-    /// `boolean` - A `true` / `false` value, for if a certain criteria is met.
     Boolean(bool),
-    /// `String` - A UTF-16 string, such as `"Hello, world"`.
-    String(JsString),
-    /// `Number` - A 64-bit floating point number, such as `3.1415`
-    Rational(f64),
-    /// `Number` - A 32-bit integer, such as `42`.
-    Integer(i32),
-    /// `BigInt` - holds any arbitrary large signed integer.
+    Float64(f64),
+    Integer32(i32),
     BigInt(JsBigInt),
-    /// `Object` - An object, such as `Math`, represented by a binary tree of string keys to Javascript values.
-    Object(JsObject),
-    /// `Symbol` - A Symbol Primitive type.
+    String(JsString),
     Symbol(JsSymbol),
+    Object(JsObject),
 }
 
-unsafe impl Trace for JsValue {
+#[allow(unsafe_op_in_unsafe_fn)]
+unsafe impl Trace for InnerValue {
     custom_trace! {this, mark, {
         if let Self::Object(o) = this {
             mark(o);
         }
     }}
 }
 
+/// A Javascript value
```

Comment:
Not blocking, but we probably should have a larger doc comment here, showing how to construct a `JsValue`.